File: src/lanelet2_map_tile_creator.cpp
//
// Created by ataparlar on 01.02.2024.
//

#include "lanelet2_map_tile_creator/lanelet2_map_tile_creator.hpp"

//#include "osmium/

#include "lanelet2_io/Io.h"
#include "lanelet2_projection/UTM.h"
#include "ogrsf_frmts.h"
#include "osmium/io/writer.hpp"
#include "proj.h"

#include <GeographicLib/MGRS.hpp>
#include <nlohmann/json.hpp>

#include <ogr_feature.h>

#include <fstream>

namespace lanelet2_map_tile_creator
{

Lanelet2MapTileCreator::Lanelet2MapTileCreator(const rclcpp::NodeOptions & options)
: Node("dynamic_lanelet2_config_metadata_creator", options)
{
  this->declare_parameter("mgrs_grid", "35TPF");
  this->declare_parameter("lanelet2_map_path", "/");
  this->declare_parameter("grid_edge_size", 5000.0);
  this->declare_parameter("osmium_config_file_path", "/");
  this->declare_parameter("metadata_file_path", "/");
  this->declare_parameter("osmium_extract_dir", "/");
  mgrs_grid = this->get_parameter("mgrs_grid").as_string();
  lanelet2_map_path = this->get_parameter("lanelet2_map_path").as_string();
  grid_edge_size = this->get_parameter("grid_edge_size").as_double();
  osmium_config_file_path = this->get_parameter("osmium_config_file_path").as_string();
  metadata_file_path = this->get_parameter("metadata_file_path").as_string();
  osmium_extract_dir = this->get_parameter("osmium_extract_dir").as_string();

  const char * pszDriverName = "GPKG";
  GDALDriver * poDriver;
  GDALAllRegister();
  poDriver = GetGDALDriverManager()->GetDriverByName(pszDriverName);
  if (poDriver == NULL) {
    printf("%s driver not available.\n", pszDriverName);
    exit(1);
  }


  std::string test = osmium_extract_dir + "output_layers.gpkg";
  GDALDataset * poDS;
  poDS = poDriver->Create(
    test.c_str(), 0, 0, 0,
    GDT_Unknown, NULL);
  if (poDS == NULL) {
    printf("Creation of output file failed.\n");
    exit(1);
  }

//  --------------------------------------------------------------------------------
  // create layer for all grids.
  OGRLayer * gridLayer;
  gridLayer = poDS->CreateLayer("grids", NULL, wkbPolygon, NULL);
  if (gridLayer == NULL) {
    printf("Layer creation failed.\n");
    exit(1);
  }

  int zone, prec;
  bool northp;
  double origin_x, origin_y, x, y;
  GeographicLib::MGRS::Reverse(mgrs_grid, zone, northp, origin_x, origin_y, prec);
  x = origin_x - 50000;
  y = origin_y - 50000;

  nlohmann::json extracts = nlohmann::json::array();
  double grid_count = 100000 / grid_edge_size;
  for (int i = 0; i < grid_count; i++) {
    for (int j = 0; j < grid_count; j++) {
      OGRFeature * gridFeature;

      gridFeature = OGRFeature::CreateFeature(gridLayer->GetLayerDefn());

      OGRPoint pt1;
      double pt1_x = x + (i * grid_edge_size);
      double pt1_y = y + (j * grid_edge_size);
      double pt1_lat;
      double pt1_lon;
      GeographicLib::UTMUPS::Reverse(zone, northp, pt1_x, pt1_y, pt1_lat, pt1_lon);
      pt1.setX(pt1_lon);
      pt1.setY(pt1_lat);

      OGRPoint pt2;
      double pt2_x = x + (i * grid_edge_size) + grid_edge_size;
      double pt2_y = y + (j * grid_edge_size);
      double pt2_lat;
      double pt2_lon;
      GeographicLib::UTMUPS::Reverse(zone, northp, pt2_x, pt2_y, pt2_lat, pt2_lon);
      pt2.setX(pt2_lon);
      pt2.setY(pt2_lat);

      OGRPoint pt3;
      double pt3_x = x + (i * grid_edge_size) + grid_edge_size;
      double pt3_y = y + (j * grid_edge_size) + grid_edge_size;
      double pt3_lat;
      double pt3_lon;
      GeographicLib::UTMUPS::Reverse(zone, northp, pt3_x, pt3_y, pt3_lat, pt3_lon);
      pt3.setX(pt3_lon);
      pt3.setY(pt3_lat);

      OGRPoint pt4;
      double pt4_x = x + (i * grid_edge_size);
      double pt4_y = y + (j * grid_edge_size) + grid_edge_size;
      double pt4_lat;
      double pt4_lon;
      GeographicLib::UTMUPS::Reverse(zone, northp, pt4_x, pt4_y, pt4_lat, pt4_lon);
      pt4.setX(pt4_lon);
      pt4.setY(pt4_lat);

      OGRPoint pt5;
      double pt5_x = x + (i * grid_edge_size);
      double pt5_y = y + (j * grid_edge_size);
      double pt5_lat;
      double pt5_lon;
      GeographicLib::UTMUPS::Reverse(zone, northp, pt5_x, pt5_y, pt5_lat, pt5_lon);
      pt5.setX(pt5_lon);
      pt5.setY(pt5_lat);

      OGRLinearRing ring;
      ring.addPoint(&pt1);
      ring.addPoint(&pt2);
      ring.addPoint(&pt3);
      ring.addPoint(&pt4);
      ring.addPoint(&pt5);

      OGRPolygon polygon;
      polygon.addRing(&ring);

      gridFeature->SetGeometry(&polygon);

      if (gridLayer->CreateFeature(gridFeature) != OGRERR_NONE) {
        printf("Failed to create feature in shapefile.\n");
        exit(1);
      }
      OGRFeature::DestroyFeature(gridFeature);
    }
  }

//  --------------------------------------------------------------------------------







  // create new layer for filtered grids
  OGRLayer * new_gridLayer;
  new_gridLayer = poDS->CreateLayer("new_grid_layer", nullptr, wkbPolygon, nullptr);
  if (new_gridLayer == NULL) {
    printf("Layer creation failed.\n");
    exit(1);
  }



  // Read lanelet2.osm
  GDALDataset * poDS_lanelet2;
  poDS_lanelet2 =
    (GDALDataset *)GDALOpenEx(lanelet2_map_path.c_str(), GDAL_OF_VECTOR, NULL, NULL, NULL);
  if (poDS_lanelet2 == NULL) {
    printf("Open failed.\n");
    exit(1);
  }
  OGRLayer * line_layer_lanelet2 = poDS_lanelet2->GetLayerByName("lines");

  // copy the lanelet2 layer from .osm file to new .gpkg file
  poDS->CopyLayer(line_layer_lanelet2, "lanelet2_map");
  OGRLayer * lanelet2_layer = poDS->GetLayerByName("lanelet2_map");

  OGRFeature * poFeature_lanelet2;
  OGRMultiLineString linestring_lanelet2_whole;


  while ((poFeature_lanelet2 = lanelet2_layer->GetNextFeature()) != NULL) {
    OGRGeometry * geometry_lanelet2 = poFeature_lanelet2->GetGeometryRef();
    OGRLineString * linestring_lanelet2 = geometry_lanelet2->toLineString();

    linestring_lanelet2_whole.addGeometry(linestring_lanelet2);
//    linestring_lanelet2_whole.get
  }


  // filter the grids with lanelet2 map area
  gridLayer->SetSpatialFilter(&linestring_lanelet2_whole);
//  linestring_lanelet2_whole.getGeometryName()


  std::ofstream metadata(metadata_file_path);

  for (auto & feature : gridLayer) {
    OGRFeature * new_feature;
    new_feature = OGRFeature::CreateFeature(gridLayer->GetLayerDefn());

    new_feature->SetGeometry(feature->GetGeometryRef());

    if (new_gridLayer->CreateFeature(new_feature) != OGRERR_NONE) {
      printf("Failed to create lanelet2_feature in shapefile.\n");
      exit(1);
    }




//    for config.json and metadata.yaml
//    --------------------------------------------------------------------------------
    nlohmann::json polygon_json = nlohmann::json::array({});
    std::string metadata_object;
    metadata_object += std::to_string(feature->GetFID()) + ".osm:\n";
    metadata_object += "  id: " + std::to_string(feature->GetFID()) + "\n";
    metadata_object += "  mgrs_code: \"" + mgrs_grid + "\"\n";

    OGRGeometry * geometry = feature->GetGeometryRef();
    OGRPolygon * polygon = geometry->toPolygon();
    bool first_point_of_polygon = true;
    for (OGRLinearRing * linearring : polygon) {
      for (const OGRPoint & point : linearring) {
        nlohmann::json point_json = nlohmann::json::array({point.getX(), point.getY()});
        polygon_json.push_back(point_json);

        if (first_point_of_polygon) {
          metadata_object += "  origin_lat: " + std::to_string(point.getY()) + "\n";
          metadata_object += "  origin_lon: " + std::to_string(point.getX()) + "\n";
          first_point_of_polygon = false;
        }
      }
    }
    metadata << metadata_object;
    nlohmann::json polygon_json2 = nlohmann::json::array({});
    polygon_json2.push_back(polygon_json);
    std::string map_name = std::to_string(feature->GetFID()) + ".osm";
    nlohmann::json extract({
      {"output", map_name},
      {"output_format", "osm"},
      {"description", "optional description"},
      {"polygon", polygon_json2},
    });
    extracts.push_back(extract);
    //    --------------------------------------------------------------------------------


  }

  poDS->DeleteLayer(0);  // delete big gridLayer
  OGRFeature::DestroyFeature(poFeature_lanelet2);

  nlohmann::json j;
  j["extracts"] = extracts;
  j["directory"] = osmium_extract_dir;
  std::ofstream o(osmium_config_file_path);
  o << std::setw(4) << j << std::endl;

  rclcpp::shutdown();
}

}  // namespace lanelet2_map_tile_creator

#include "rclcpp_components/register_node_macro.hpp"
RCLCPP_COMPONENTS_REGISTER_NODE(lanelet2_map_tile_creator::Lanelet2MapTileCreator)

File: scripts/lanelet2_map_tile_creator.py
import sys
import mgrs
import utm
from osgeo import gdal
from osgeo import ogr
import yaml
import json

if __name__ == "__main__":

    mgrs_grid = "35TPF"
    grid_edge_size = 50
    lanelet2_map_path = "/home/ataparlar/data/gis_data/lanelet2_map_tile_creator/maps/lanelet2_map_sorted.osm"
    metadata_file_path = "/home/ataparlar/data/gis_data/lanelet2_map_tile_creator/test/metadata.yaml"
    osmium_config_file_path = "/home/ataparlar/data/gis_data/lanelet2_map_tile_creator/test/test_config.json"
    osmium_extract_dir = "/home/ataparlar/data/gis_data/lanelet2_map_tile_creator/test/"

    mgrs_object = mgrs.MGRS()
    zone, northp, y, x = mgrs_object.MGRSToUTM(mgrs_grid)
    print("x: ", x)
    print("y: ", y)
    print("zone: ", zone)
    print("northp: ", northp)

    # Create json
    # --------------------------------------------------------------------------------

    driverName = "GPKG"
    drv = gdal.GetDriverByName(driverName)
    if drv is None:
        print("%s driver not available.\n" % driverName)
        sys.exit(1)

    ds_grids = drv.Create(osmium_extract_dir + "output_layers.gpkg", 0, 0, 0, gdal.GDT_Unknown)
    if ds_grids is None:
        print("Creation of output file failed.\n")
        sys.exit(1)

    layer_grids = ds_grids.CreateLayer("grids", None, ogr.wkbPolygon)
    if layer_grids is None:
        print("Layer creation failed.\n")
        sys.exit(1)

    layer_lanelet2_whole = ds_grids.CreateLayer("lanelet2_whole", None, ogr.wkbMultiLineString)
    if layer_lanelet2_whole is None:
        print("Layer creation failed layer_lanelet2_whole.\n")
        sys.exit(1)

    # Create Grid Layer
    # --------------------------------------------------------------------------------
    grid_count = 100000 / grid_edge_size
    for i in range(int(grid_count)):
        print(i)
        for j in range(int(grid_count)):
            feature_grid = ogr.Feature(layer_grids.GetLayerDefn())

            linear_ring = ogr.Geometry(ogr.wkbLinearRing)

            for a in range(5):
                pt_x, pt_y = 0.0, 0.0
                if (a == 0) or (a == 4):
                    pt_x = x + (i * grid_edge_size)
                    pt_y = y + (j * grid_edge_size)
                elif a == 1:
                    pt_x = x + (i * grid_edge_size) + grid_edge_size
                    pt_y = y + (j * grid_edge_size)
                elif a == 2:
                    pt_x = x + (i * grid_edge_size) + grid_edge_size
                    pt_y = y + (j * grid_edge_size) + grid_edge_size
                elif a == 3:
                    pt_x = x + (i * grid_edge_size)
                    pt_y = y + (j * grid_edge_size) + grid_edge_size
                pt_lat, pt_lon = utm.to_latlon(pt_y, pt_x, zone, northp)
                linear_ring.AddPoint_2D(pt_lon, pt_lat)

            polygon = ogr.Geometry(ogr.wkbPolygon)
            polygon.AddGeometry(linear_ring)

            feature_grid.SetGeometry(polygon)
            if layer_grids.CreateFeature(feature_grid) != 0:
                print("Failed to create feature in shapefile.\n")
                sys.exit(1)

            feature_grid.Destroy()
    # --------------------------------------------------------------------------------

    # Create new layer for filtered grids
    # --------------------------------------------------------------------------------
    layer_filtered_grids = ds_grids.CreateLayer("filtered_grids", None, ogr.wkbPolygon)
    if layer_filtered_grids is None:
        print("Layer creation failed.\n")
        sys.exit(1)

    # Read Lanelet2 Map
    # --------------------------------------------------------------------------------
    ds_lanelet2 = gdal.OpenEx(lanelet2_map_path, gdal.OF_VECTOR)
    if ds_lanelet2 is None:
        print("Open failed.\n")
        sys.exit(1)

    # Copy the lanelet2 lines layer to export layer
    # --------------------------------------------------------------------------------
    ds_grids.CopyLayer(ds_lanelet2.GetLayerByName("lines"), "lanelet2_map")
    layer_lanelet2 = ds_grids.GetLayerByName("lanelet2_map")

    # Make a multilinestring in order to use in filtering
    # --------------------------------------------------------------------------------
    lanelet2_whole_mls = ogr.Geometry(ogr.wkbMultiLineString)
    for feature_lanelet2_linestring in layer_lanelet2:
        lanelet2_linestring = feature_lanelet2_linestring.GetGeometryRef()
        lanelet2_whole_mls.AddGeometry(lanelet2_linestring)

    feature_lanelet2_whole = ogr.Feature(layer_lanelet2_whole.GetLayerDefn())
    feature_lanelet2_whole.SetGeometry(lanelet2_whole_mls)
    if layer_lanelet2_whole.CreateFeature(feature_lanelet2_whole) != 0:
        print("Failed to create feature in shapefile.\n")
        sys.exit(1)

    # Filter and destroy feature
    # --------------------------------------------------------------------------------
    layer_grids.SetSpatialFilter(lanelet2_whole_mls)
    print("layer_grids.GetFeatureCount(): ", layer_grids.GetFeatureCount())
    feature_lanelet2_whole.Destroy()

    # Set filtered grid layer
    # --------------------------------------------------------------------------------
    for grid in layer_grids:
        geometry_grid = grid.GetGeometryRef()

        filtered_feature_grid = ogr.Feature(layer_filtered_grids.GetLayerDefn())
        filtered_feature_grid.SetGeometry(geometry_grid)
        if layer_filtered_grids.CreateFeature(filtered_feature_grid) != 0:
            print("Failed to create feature in shapefile.\n")
            sys.exit(1)

        filtered_feature_grid.Destroy()

    # Create yaml data and write to file
    # --------------------------------------------------------------------------------
    metadata_yaml = {}
    for filtered_grid in layer_filtered_grids:
        geometry_filtered_grid = filtered_grid.GetGeometryRef()
        point_lat = 0.0
        point_lon = 0.0
        for linearring in geometry_filtered_grid:
            point_lat = linearring.GetPoint(0)[0]
            point_lon = linearring.GetPoint(0)[1]
        file_id = str(filtered_grid.GetFID()) + ".osm"
        yaml_data = {
            file_id: {
                "mgrs_code": mgrs_grid,
                "origin_lat": point_lat,
                "origin_lon": point_lon
            }
        }
        metadata_yaml.update(yaml_data)
    yaml_print = yaml.dump(metadata_yaml, sort_keys=False)

    with open(metadata_file_path, 'w', ) as f:
        yaml.dump(metadata_yaml, f, sort_keys=False)


    # Create config.json for Osmium Extract
    # --------------------------------------------------------------------------------
    extracts = []
    for filtered_grid in layer_filtered_grids:
        polygon_inner = []
        geometry_filtered_grid = filtered_grid.GetGeometryRef()
        print(geometry_filtered_grid)

        # print("geometry_grid.GetGeometryCount(): ", geometry_filtered_grid.GetGeometryCount())

        # test_polygon = ogr.Geometry(ogr.wkbPolygon)
        # test_polygon.AddPoint()

        counter = 0
        for linearring in geometry_filtered_grid:
            for i in range(5):
                point_lat = linearring.GetPoint(i)[1]
                point_lon = linearring.GetPoint(i)[0]
                point_list = [point_lon, point_lat]
                polygon_inner.append(point_list)



        polygon_outer = [polygon_inner]
        extract_element = {
            "description" : "optional description",
            "output" : str(filtered_grid.GetFID()) + ".osm",
            "output_format" : "osm",
            "polygon" : polygon_outer
        }
        extracts.append(extract_element)

    config_json_ = {
        "directory" : osmium_extract_dir,
        "extracts" : extracts
    }

    config_json = json.dumps(config_json_, indent=2)

    with open(osmium_config_file_path, "w") as write_file:
        write_file.write(config_json)








